federatedscope/core/trainers/context.py

```diff
@@ -117,51 +152,107 @@ def __init__(self,
 
         self.lifecycles = collections.defaultdict(set)
 
-        if init_attr:
-            # setup static variables for training/evaluation
-            self.setup_vars()
-
-    def setup_vars(self):
+        # Setup optimize-related context variable
         if self.cfg.backend == 'torch':
             self.trainable_para_names = get_trainable_para_names(self.model)
+            # TODO: make `criterion` and `regularizer` @property and cached
+            #  to compare whether changes happen
             self.criterion = get_criterion(self.cfg.criterion.type,
                                            self.device)
             self.regularizer = get_regularizer(self.cfg.regularizer.type)
             self.grad_clip = self.cfg.grad.grad_clip
-            if isinstance(self.data, ClientData):
-                self.data.setup(self.cfg)
         elif self.cfg.backend == 'tensorflow':
             self.trainable_para_names = self.model.trainable_variables()
             self.criterion = None
             self.regularizer = None
             self.optimizer = None
             self.grad_clip = None
 
-        # Process training data
-        if self.get('train_data', None) is not None or self.get(
-                'train_loader', None) is not None:
-            # Calculate the number of update steps during training given the
-            # local_update_steps
-            self.num_train_batch, self.num_train_batch_last_epoch, \
-                self.num_train_epoch, self.num_total_train_batch = \
+    # Train related property, query from `cfg` if not set
+    @property
+    def num_train_batch(self):
+        if self.get('num_train_batch'):
+            return self.get('num_train_batch')
+        return self._calculate_batch_epoch_num(mode='train')[0]
```

Comment:
The number of epoch is decided by the selected dataset. Maybe we should name the parameter as `split` or `dataset` here rather than `mode`